core/src/test/java/io/questdb/test/griffin/wal/SnapshotFuzzTest.java

```diff
@@ -0,0 +1,215 @@
+                0,
+                1,
+                0,
+                0.5,
+                0
+        );
+
+        fuzzer.setFuzzCounts(
+                rnd.nextBoolean(),
+                rnd.nextInt(2_000_000),
+                rnd.nextInt(1000),
+                rnd.nextInt(3),
+                rnd.nextInt(5),
+                rnd.nextInt(1000),
+                rnd.nextInt(1_000_000),
+                5 + rnd.nextInt(10)
+        );
+
+        setFuzzProperties(1, getRndO3PartitionSplit(rnd), getRndO3PartitionSplitMaxCount(rnd), 10 * Numbers.SIZE_1MB, 3);
+        runFuzzWithSnapshot(rnd);
+    }
+
+    private void createSnapshot() throws SqlException {
+        setProperty(PropertyKey.CAIRO_SNAPSHOT_INSTANCE_ID, "id_1");
+        LOG.info().$("starting snapshot").$();
+
+        ddl("snapshot prepare");
+        CairoConfiguration conf = engine.getConfiguration();
+
+        FilesFacade ff = conf.getFilesFacade();
+        Path snapshotPath = Path.getThreadLocal(conf.getRoot()).put("_snapshot").$();
+        Path rootPath = Path.getThreadLocal2(conf.getRoot()).$();
+
+        ff.mkdirs(snapshotPath, conf.getMkDirMode());
+
+        LOG.info().$("copying data to snapshot [from=").$(rootPath).$(", to=").$(snapshotPath).$();
+        ff.copyRecursive(rootPath, snapshotPath, conf.getMkDirMode());
+
+        ddl("snapshot complete");
+    }
+
+    private void fullFuzz(Rnd rnd) {
+        fuzzer.setFuzzProbabilities(
+                0.5 * rnd.nextDouble(),
+                rnd.nextDouble(),
+                rnd.nextDouble(),
+                0.5 * rnd.nextDouble(),
+                rnd.nextDouble(),
+                rnd.nextDouble(),
+                rnd.nextDouble(),
+                rnd.nextDouble(),
+                0.1 * rnd.nextDouble(), 0.01,
+                rnd.nextDouble()
+        );
+
+        fuzzer.setFuzzCounts(
+                rnd.nextBoolean(),
+                rnd.nextInt(2_000_000),
+                rnd.nextInt(1000),
+                rnd.nextInt(3),
+                rnd.nextInt(5),
+                rnd.nextInt(1000),
+                rnd.nextInt(1_000_000),
+                5 + rnd.nextInt(10)
+        );
+    }
+
+    private void restoreSnapshot() {
+        LOG.info().$("begin snapshot restore").$();
+        engine.releaseInactive();
+
+        CairoConfiguration conf = engine.getConfiguration();
+        FilesFacade ff = conf.getFilesFacade();
+        Path snapshotPath = Path.getThreadLocal(conf.getRoot()).put("_snapshot").slash$();
+        Path rootPath = Path.getThreadLocal2(conf.getRoot()).slash$();
+
+        ff.rmdir(rootPath);
+        ff.rename(snapshotPath, rootPath);
+
+        setProperty(PropertyKey.CAIRO_SNAPSHOT_INSTANCE_ID, "id_2");
+
+        LOG.info().$("recovering from snapshot").$();
+        engine.recoverSnapshot();
+        engine.getTableSequencerAPI().releaseAll();
+        engine.reloadTableNames();
+    }
+
+    protected void runFuzzWithSnapshot(Rnd rnd) throws Exception {
```

Comment:
nit: might be a good idea to run a server instance with PGWire/ILP interface and then generate writes and take a snapshot concurrently. There is a chance we might find more bugs around snapshots this way.

Reply:
So far I found places that break append-only or copy-on-write storage assumptions to be a problem for snapshots. These storage level fuzz tests exercise a pretty good combination of storage actions we can do, much more than the ILP fuzz framework we have.

PgWire/ILP cannot do storage modifications not covered in these tests and we don't have a good transaction fuzzing framework for them, so it is relatively expensive to add such tests.